### PR TITLE
Feature: FPP Visualize

### PR DESCRIPTION
**Pull request by thomas-bc**
| | |
|:---|:---|
|**_Originating Project/Creator_**| |
|**_Affected Component_**|  |
|**_Affected Architectures(s)_**|  |
|**_Related Issue(s)_**| https://github.com/nasa/fprime/issues/2059 |
|**_Has Unit Tests (y/n)_**|  |
|**_Builds Without Errors (y/n)_**|  |
|**_Unit Tests Pass (y/n)_**|  |
|**_Documentation Included (y/n)_**|  |

---
## Change Description

**Depends on**:
- https://github.com/fprime-community/fprime-visual/pull/1
- https://github.com/nasa/fprime/issues/2084

Introducing: `fprime-util visualize`
Within a Topology directory, performs the following

-  fpp-to-xml
-  fpl-convert-xml
-  fpl-layout
-  start fprime-visual Flask app to serve visualization

Effectively, this allows with a single command to spin up a visualization of the topology and all its subtopology graphs.

```bash
cd Ref/Top
fprime-util visualize
```

## Rationale

Major QoL improvement to visualization capabilities. This is currently doable, but requires a lot of manual installation of tools etc...

## Todo
- The 2 things listed in _Depends on_
- Probably add some tests

## Note

Opening as draft in case there's change requests. Do not merge for now, there are a couple of dependencies for this that need to be released before this can go through.

### CODE REVIEW

**Comment by thomas-bc** [comment]
Fails tests because `fprime_visual` pip package is not currently published.

File: src/fprime/fpp/visualize.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    Args:
+        build: build directory output
+        parsed: parsed input arguments
+        _: unused cmake_args
+        __: unused make_args
+        ___: unused pass-through arguments
+    """
+
+    # We could use the build directory, but it's quirky because not managed by CMake ??
```
Comment (thomas-bc) [inline src/fprime/fpp/visualize.py]: Thougts?

File: src/fprime/fpp/visualize.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    xml_cache = Path(".fpp-viz-cache/xml")
+    xml_cache.mkdir(parents=True, exist_ok=True)
+
+    # Run fpp-to-xml
+    FppUtility("fpp-to-xml").execute(
+        build,
+        parsed.path,
+        args=(
+            {},
+            ["--directory", ".fpp-viz-cache/xml"],
```
Comment (thomas-bc) [inline src/fprime/fpp/visualize.py]: Need rework so `--directory` influences the top level `.fpp-viz-cache` only and everything flows down from it.

File: src/fprime/fpp/visualize.py (hunk 1 above)
Comment (thomas-bc) [reply src/fprime/fpp/visualize.py]: Using `.visualize-cache/` in current directory for now, let me know if you think we should change that.

**Comment by thomas-bc** [comment]
Ready for review, as all the pieces are there, see https://github.com/nasa/fprime/issues/2059#issuecomment-1622704571

File: src/fprime/fpp/cli.py [hunk 3]
```diff
@@ -35,6 +35,31 @@ def run_fpp_check(
...
 
 
+def run_fpp_to_xml(
+    build: "Build",
+    parsed: argparse.Namespace,
+    _: Dict[str, str],
+    __: Dict[str, str],
+    ___: List[str],
+):
+    """Run fpp check application
```
Comment (LeStarch) [inline src/fprime/fpp/cli.py]: Comment out of date.  Should this not say something about fpp-to-xml.

File: src/fprime/fpp/cli.py [hunk 4]
```diff
@@ -56,4 +81,11 @@ def add_fpp_parsers(
     check_parser.add_argument(
         "-u", "--unconnected", default=None, help="write unconnected ports to file"
     )
-    return {"fpp-check": run_fpp_check}, {"fpp-check": check_parser}
+    fpp_to_xml_parser = subparsers.add_parser(
+        "fpp-to-xml", help="Runs fpp-to-xml utility", parents=[common], add_help=False
```
Comment (LeStarch) [inline src/fprime/fpp/cli.py:83]: Does `fpp-to-xml` need to be a public thing?  Could it not just be an internal function?  Given that we intend to remove XML it seems like maybe we should just internalize this.

File: src/fprime/fpp/visualize.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        "--gui-port",
+        help="Set the GUI port for fprime-visual [default: %(default)s]",
+        required=False,
+        default=7000,
+    )
+    viz_parser.add_argument(
+        "--cache-dir",
+        help="Set the directory to store layout files in [default: %(default)s]",
+        required=False,
+        default=".visualize-cache",
```
Comment (LeStarch) [inline src/fprime/fpp/visualize.py]: Should the default be in `/tmp`?  I don't see reason to keep it around locally.

**Review by LeStarch — COMMENTED** [review]
A few questions, but looks generally good.

File: src/fprime/fpp/visualize.py (hunk 5 above)
Comment (thomas-bc) [reply src/fprime/fpp/visualize.py]: True - if we expose the option to change where it goes, might as well make the default `/tmp`. Will change

File: src/fprime/fpp/cli.py (hunk 4 above)
Comment (thomas-bc) [reply src/fprime/fpp/cli.py:83]: Discussed internally, we will keep fpp-to-xml public for now

File: src/fprime/fpp/visualize.py (hunk 1 above)
Comment (thomas-bc) [reply src/fprime/fpp/visualize.py]: Changed to `/tmp/fprime-visualize`

File: src/fprime/fpp/visualize.py [hunk 6]
```diff
@@ -146,9 +146,9 @@ def add_fpp_viz_parsers(
         default=7000,
     )
     viz_parser.add_argument(
-        "--cache-dir",
+        "--working-dir",
         help="Set the directory to store layout files in [default: %(default)s]",
         required=False,
-        default=".visualize-cache",
+        default="/tmp/fprime-visualize",
```
Comment (LeStarch) [inline src/fprime/fpp/visualize.py]: What happens when this directory does not give write permissions?  Do we tell the user that they should override with the `--working-dir` flag.

**Review by LeStarch — DISMISSED** [review]
I had one more minor question, but not sure it should trip up the PR.

File: src/fprime/fpp/visualize.py (hunk 6 above)
Comment (thomas-bc) [reply src/fprime/fpp/visualize.py]: Good point, I added a more verbose error with resolution approach.

**Review by LeStarch — APPROVED** [review]

**Pull request merged**